exonum/src/storage/sparse_list_index.rs

```diff
@@ -0,0 +1,838 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.push(1);
+    /// assert_eq!(Some(1), index.get(0));
+    ///
+    /// index.set(0, 10);
+    /// assert_eq!(Some(10), index.get(0));
+    /// ```
+    pub fn set(&mut self, index: u64, value: V) {
+        let mut size = self.size();
+        // Update items count
+        if self.base.get::<u64, V>(&index).is_none() {
+            size.length += 1;
+            if index >= size.capacity {
+                size.capacity = index + 1;
+            }
+            self.set_size(size);
+        }
+        self.base.put(&index, value)
+    }
+
+    /// Clears the list, removing all values.
+    ///
+    /// # Notes
+    ///
+    /// Currently this method is not optimized to delete a large set of data.
+    /// During the execution of this method the amount of allocated memory
+    /// is linearly dependent on the number of elements
+    /// in the index.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.push(1);
+    /// assert!(!index.is_empty());
+    ///
+    /// index.clear();
+    /// assert!(index.is_empty());
+    /// ```
+    pub fn clear(&mut self) {
+        self.size.set(Some(SparseListSize::default()));
+        self.base.clear()
+    }
+
+    /// Appends an element to the back of the 'SparceListIndex'.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.push_back(1);
+    /// assert!(!index.is_empty());
+    /// ```
+    pub fn push_back(&mut self, value: V) {
+        self.push(value);
+    }
+
+    /// Removes the first element from the 'SparceListIndex' and returns it, or None if it is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert_eq!(None, index.pop());
+    ///
+    /// index.push_back(1);
+    /// assert_eq!(Some(1), index.pop_front());
+    /// ```
+    pub fn pop_front(&mut self) -> Option<V> {
+        let _opt = {
+            self.indices().next()
+        };
+        if let Some(min_index) = _opt {
+            return self.remove(min_index);
```

Comment:
Сall `remove` method is formally correct here, but it is not effective because `remove` does additional checks that are not necessary in this case:

- We are guaranteed to know that the index is less than capacity
- We are guaranteed to know that an element with such an index exists

And since we still need get the value from the index, you may use `self.iter().next()` instead of `self.indices().next()`